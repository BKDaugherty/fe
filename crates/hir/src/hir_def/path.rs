use crate::hir_def::Partial;

use super::IdentId;

#[salsa::interned]
pub struct PathId {
    pub data: Vec<Partial<IdentId>>,
}
