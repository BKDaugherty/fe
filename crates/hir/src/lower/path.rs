use parser::{ast, SyntaxToken};

use crate::hir_def::{IdentId, Partial, PathId, PathSegment};

use super::FileLowerCtxt;

impl PathId {
    pub(super) fn lower_ast(ctxt: &mut FileLowerCtxt<'_>, ast: ast::Path) -> Self {
        let mut segments = Vec::new();
        for seg in ast.into_iter() {
            let segment = match seg.kind() {
                Some(ast::PathSegmentKind::Ingot(_)) => Some(PathSegment::Ingot),
                Some(ast::PathSegmentKind::Super(_)) => Some(PathSegment::Super),
                Some(ast::PathSegmentKind::SelfTy(_)) => Some(PathSegment::SelfTy),
                Some(ast::PathSegmentKind::Self_(_)) => Some(PathSegment::Self_),
                Some(ast::PathSegmentKind::Ident(ident)) => {
                    Some(PathSegment::Ident(IdentId::lower_token(ctxt, ident)))
                }
                None => None,
            }
            .into();
            segments.push(segment);
        }

        Self::new(ctxt.db, segments)
    }

    pub(super) fn lower_ast_partial(
        ctxt: &mut FileLowerCtxt<'_>,
        ast: Option<ast::Path>,
    ) -> Partial<Self> {
        ast.map(|ast| Self::lower_ast(ctxt, ast)).into()
    }

    pub(super) fn from_ident(ctxt: &mut FileLowerCtxt<'_>, ast: SyntaxToken) -> Self {
        let ident_id = IdentId::new(ctxt.db, ast.text().to_string());
        let seg = vec![Partial::Present(PathSegment::Ident(ident_id))];
        Self::new(ctxt.db, seg)
    }
}
