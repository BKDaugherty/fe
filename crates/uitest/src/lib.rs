use std::path::Path;

use dir_test::{dir_test, Fixture};
use driver::DriverDataBase;
use fe_compiler_test_utils::snap_test;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/fixtures/name_resolution",
    glob: "*.fe"
)]
#[allow(unused)]
fn run_name_resolution(fixture: Fixture<&str>) {
    run_name_resolution_impl(fixture);
}

fn run_name_resolution_impl(fixture: Fixture<&str>) {
    let mut driver = DriverDataBase::default();
    let path = Path::new(fixture.path());
    driver.run_on_file(path);
    let diags = driver.format_diags();
    snap_test!(diags, fixture.path());
}

#[cfg(target_family = "wasm")]
mod wasm {
    use super::*;
    use wasm_bindgen_test::wasm_bindgen_test;

    #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/fixtures/name_resolution",
        glob: "*.fe",
    )]
    #[dir_test_attr(#[wasm_bindgen_test])]
    fn run_name_resolution(fixture: Fixture<&str>) {
        run_name_resolution_impl(fixture);
    }
}
