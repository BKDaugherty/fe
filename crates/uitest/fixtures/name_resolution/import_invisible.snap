---
source: crates/uitest/src/lib.rs
expression: diags
input_file: crates/uitest/fixtures/name_resolution/import_invisible.fe
---
error[2-0003]: `Bar` is not visible
  ┌─ import_invisible.fe:1:10
  │
1 │ use foo::Bar
  │          ^^^ `Bar` is not visible
  ·
4 │     struct Bar {}
  │            --- `Bar is defined here


